### DISCUSSION TITLE
Trying to set hostname in tab_title_template

### DISCUSSION THREAD

**Discussion by nst0022** [kovidgoyal/kitty#3196] [Q&A] [answered]
There are some predefined fields for the "tab_title_template" option, like "index", "title" (= user name), asf.

I was trying to get the hostname somehow entered, but failed.

As I can write `{title.upper()}`, I tried `{os.usname()[1]}`, but to no avail.

When I try that in Python with an f-string, it works, if I make the appropriate import beforehand:

```
>>> from os import uname
>>> print(f'{uname()[1]}')
xyz
```

So, it would probably work in kitty.conf as well, if I could somehow use an import instruction.

Or can I specify a kitten in the title?

I have also tried `{$(hostname)}` as I saw in Kovid's reply in the Welcome tread (there: `include hostname-${HOSTNAME}.conf`), hoping this parameter is processed by a shell as well, but no.

**Comment by kovidgoyal** [comment]
Whats the use ase? If you use os.usname() in a template all tabs will get the same hostname so what's the point? I am guessing you want the tab title to show the hostname when you ssh into some host in the active window in that tab. The way to do that is to set the window title or tab title using either escape codes or kitty remote control before sshing in. So write a wrapper script for ssh to do that, and use it to ssh.

**Reply by nst0022** [reply]
OK, I understand the wrapper script for ssh would set the tab title, but when I come back, I can probably not trigger a second script that restores it to its previous state.

So, bad idea in the first place :-).

**Comment by kovidgoyal** [comment]
On Tue, Dec 29, 2020 at 06:31:45AM -0800, nst0022 wrote:
> OK, I understand the wrapper script for ssh would set the tab title, but when I come back, I can probably not trigger a second script that restores it to its previous state.

Sure you can:

#!/bin/sh
kitty @ set-tab-title whatever
ssh whatever
kitty @ set-tab-title

**Comment by nst0022** [comment]
I had to go a step back and figure out, how tab titles and window titles generally work, and noticed two peculiarities:

(1) Tab titles
Say the tab title contains {title}, then the userid is displayed.
If I switch the user with "su", the new userid is displayed, as expected.
If I return from "su", the original userid is restored, again as expected.
However, if I ssh to another host and switch to another user there, the original userid remains untouched.
Bug or feature?

(2) OS window titles
If I open kitty with the --title command line option, the title is displayed in the window's title bar.
If I open any new tab, the title in the window's title bar is overwritten with the userid.
Only the very first tab shows the window's title specified in the command line parameter.
Bug or feature?

**Comment by kovidgoyal** [comment]
On Tue, Dec 29, 2020 at 10:45:16PM -0800, nst0022 wrote:
> I had to go a step back and figure out, how tab titles and window titles generally work, and noticed two peculiarities:
> 
> (1) Tab titles
> Say the tab title contains {title}, then the userid is displayed.
> If I switch the user with "su", the new userid is displayed, as expected.
> If I return from "su", the original userid is restored, again as expected.
> However, if I ssh to another host and switch to another user there, the original userid remains untouched.
> Bug or feature?

What title is displayed depends on what the program running in the
window sets the title to. Presumably, the computer you are sshing into
does not set window titles.

> 
> (2) OS window titles
> If I open kitty with the --title command line option, the title is displayed in the window's title bar.
> If I open any new tab, the title in the window's title bar is overwritten with the userid.
> Only the very first tab shows the window's title specified in the command line parameter.
> Bug or feature?


--title sets window title, not OS window title. The OS window title
simply inherits from the current active window title. Similarly tab
titles inherit from active window title, unless overriden.

**Comment by nst0022** [comment]
Sorry, the discussion gets too complicated, I have to quit it.

I have already an alternative solution to my initial intention.

No hard feelings :-).

**Reply by aorith** [reply]
I wanted that same behaviour, to do that I use this wrapper function: https://github.com/aorith/dotfiles/blob/133d30176e70758da2b90b5affc08e292d57262e/topics/shell/etc/shell/functions#L40 

That sets the title to user@hostname and is kept meanwhile you're ssh'ed if the remote doesn't override it, and resets once the connection quits, I don't really use the "reset" part since I override the title on my machines, so it gets reset to localuser@localhostname once I quit and immediately changes to whatever I have configured in my shell.

But please note that overriding TERM is NOT recommended.

**Comment by nst0022** [comment]
@aorith:

Your ssh wrapper does not work here, because the output of `ssh -G` shows an IP address as `hostname`, rather than the content of `/etc/hostname`, which can be retrieved with `/bin/hostname`, so I get `192.168.22.33` instead of `xyz`, in fact, `xyz` is nowhere else present in this large output (I tried to grep it).

Besides that, another problem would be, if I wanted to set a tab title while on the remote host, I needed to install kitty there, which would either be an additional hassle, or a show stopper. 

Anyway, thanks for your reply.

And, of course, I would never override TERM :-).

**Reply by aorith** [reply]
@nst0022, You're right! it will show the IP as it just takes the parameters and/or default username passed to ssh, as I always log via dns name I didn't even take that into account.

But you don't need to install kitty on the remote for that solution to work as it just uses escape codes sequences to set the title.

If you don't mind and find another solution, please post it here :)

Regards,

**Comment by nst0022** [comment]
> But you don't need to install kitty on the remote for that solution to work as it just uses escape codes sequences to set the title.

I was more interested in setting the tab titles, rather than the window title :-)

>If you don't mind and find another solution, please post it here :)

I will do that :-)

**Comment by nst0022** [comment, marked as answer]
Problem solved :-).

First, thanks to @aorith for pointing me to your escape sequences. This was the key to the solution. 

Second, my apologies to Kovid that I had not understood you from the beginning. I am just a few days with kitty and there is definitely some learning to do :-).

Initially a disambiguation:

I call the outermost window kitty runs in 'window' and its title the 'window title', a new tab is a 'tab' and its title the 'tab title', and any new window produced with ctrl-shift-enter a 'split window' (which has no visible title).

Third, the problem I had with overwriting the 'window title' laid in each and every `.bashrc` file I have on all my machines. After setting the PS1 prompt, way down below, I had some default code active (thinking it was passive and did not hurt) where the PS1 prompt was extended to setting the 'window title' to the userid. That was the cause why I always claimed the '{title}' equals the userid :-).

So, I replaced that part of the prompt with hostname and userid and synchronized all `.bashrc` files I have.

Fourth, I learned, that each 'split window' has its own hidden 'window title', so there is no need to place the hostname or the user on a tab explicitly (I just show the index now, to make the tab narrow, but I could also add the title). Upon selection of any 'split window', the visible 'window title' (and the respective 'tab title' if specified) is updated by kitty automatically.

So, just a clean `.bashrc` on every machine for every user does the trick :-).

This is the part, where the window title is prepended:

`PS1="\e]0;$(hostname) $(whoami)\007$PS1"`

Making this a part of the prompt ensures that each `su` and `ssh` is covered, no wrappers required.
